Swatch Treatments

dani97 opened the pull request:
<!--
Before submitting this pull request, please make sure you have read our Contribution Guidelines and your PR meets our contribution standards:
https://github.com/magento-research/pwa-studio/blob/master/.github/CONTRIBUTING.md

Please fill out as much information as you can about your PR to help speed up the review process.
If your PR addresses an existing GitHub Issue, please refer to it in the title or Additional Information section to make the connection.

We may ask you for changes in your PR in order to meet the standards set in our Contribution Guidelines. PRs that do not comply with our guidelines may be closed at the maintainers' discretion.

Feel free to remove this section before creating this PR. Thank you for your contribution!
-->

## Description

Update Swatch Treatments.

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here by replacing ISSUE_NUMBER with your actual issue number. -->
<!--- Using the above wording causes Github to automatically close the issue on merge. -->
Closes #1390 .

## Verification Steps
<!-- Please describe in detail how a reviewer can verify your changes. -->
1. Go to the configurable PDP.
2. Verify the tool-tips not shown when selecting the options and new treatment comes.
3. Make sure the same working in cart.

## Screenshots / Screen Captures (if appropriate)
![image](https://user-images.githubusercontent.com/13298685/62407732-179ad880-b5db-11e9-8c89-7b9525e9efbb.png)

## Proposed Labels for Change Type/Package
<!--- What type of change level would you suggest for this PR? -->
<!--- Major, Minor, or Patch? -->
<!--- See https://magento-research.github.io/pwa-studio/technologies/versioning/ for help -->
- [ ] major (e.g x.0.0 - a breaking change)
- [X] minor (e.g 0.x.0 - a backwards compatible addition)
- [ ] patch (e.g 0.0.x - a bug fix)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] I have updated the documentation accordingly, if necessary.
- [ ] I have added tests to cover my changes, if necessary.

sirugh commented on packages/venia-concept/src/components/ProductOptions/tile.js:
```diff
@@ -40,11 +39,11 @@ class Tile extends Component {
-                    <span>{label}</span>
-                </button>
-            </Tooltip>
+
```
Can remove whitespace.

sirugh commented on packages/venia-concept/src/components/ProductOptions/tile.js:
```diff
@@ -2,7 +2,6 @@ import React, { Component } from 'react';
 import { bool, number, shape, string } from 'prop-types';
 
 import classify from '../../classify';
-import Tooltip from './toolTip';
```
You can also delete the entire Tooltip component and related tests, I believe.

sirugh commented on packages/venia-concept/src/components/ProductOptions/swatch.js:
```diff
@@ -68,16 +67,13 @@ class Swatch extends Component {
-                    {...restProps}
-                    className={className}
-                    style={finalStyle}
-                    title={label}
```
Any reason why you removed the `title` prop?

dani97 replied:
Just to avoid default tooltip generated created by the title attribute.
I will disable with css if it affects a11y

jimbo replied:
That default tooltip from the `title` attribute is very important for accessibility, even for users without vision impairment. I do suggest restoring it (and not hiding it with CSS).

brendanfalkowski replied:
It doesn't affect this context (button with only an icon as child content), but I read somewhere that this was an accessibility anti-pattern:

`<button title="Thing">Thing</button>`

Which a screen reader would announce as: 

`BUTTON Thing TITLE Thing`

Has anyone encountered that, or has the practice changed? Years ago a screenreader user was begging people to stop putting redundant title tags on everything because it clogged the audio.

dani97 replied:
sure will restore @jimbo @brendanfalkowski

sirugh commented on packages/venia-concept/src/components/ProductOptions/option.js:
```diff
@@ -54,6 +69,9 @@ class Option extends Component {
                     onSelectionChange={handleSelectionChange}
                 />
+                <p className={classes.selection}>
+                    {selectedValue?`Selected ${label} : ${this.valuesMap.get(Array.from(selectedValue).pop())}`:''}
```
I think this would be more readable if you pulled it out above the `return`.

For example:
```js
const selectedLabel = selectedValue ? ... : '';

return (
  ...
   <p className={classes.selection>
      {selectedLabel}
  </p>
  ...
);
```

dani97 replied:
sure will make that change. :+1:

sirugh commented on packages/venia-concept/src/components/ProductOptions/option.js:
```diff
@@ -10,6 +10,20 @@ import defaultClasses from './option.css';
 
 class Option extends Component {
+
+    constructor(props) {
```
How do you feel about rewriting this component as a function component? Look at some of the rewrites we've been doing lately for reference but basically we will convert things like state to `useState` and the creation of the values map to a `useMemo` generator.

dani97 replied:
Will refactor to a function components @sirugh

sirugh commented on packages/venia-concept/src/components/ProductOptions/option.js:
```diff
@@ -54,6 +69,9 @@ class Option extends Component {
                     onSelectionChange={handleSelectionChange}
                 />
+                <p className={classes.selection}>
+                    {selectedValue?`Selected ${label} : ${this.valuesMap.get(Array.from(selectedValue).pop())}`:''}
```
I'm also a little worried about human readable label values. Will it be `Selected fashion_color` or is the values label already human readable?

dani97 replied:
I have kept the map  as { attribute_code : store_label } So the text will be `Selected Fashion Color<attribute_label>: Burgundy<store_label>`

dani97 commented:
@jimbo As per @sirugh  comment I refactored options component into function component but I could not fix  tests
1 ) does not call onSelectionChange if not provided
2 ) calls onSelectionChange function with optionId and selection
These two tests get options instance and call `handleSelectionChange` method.
Since I refactored it into function component, I cannot access the instance. Please advise on this.

jimbo commented:
@dani97 Thanks for taking this on. Those tooltips were bothering me for a long time! 😅 

I've updated the tests for you; have a look and let me know if you have any questions. Make sure to use `createTestInstance` from Peregrine, and use the [Test Renderer docs](https://reactjs.org/docs/test-renderer.html) to work with the returned instance.

I also made some minor changes to the `Option` component. In general, it's better to not keep the `selection` set in multiple components' state, so I swapped `selection` to be the derived state (the label of the selected option). Should simplify things a bit.

jimbo commented on packages/venia-concept/src/components/ProductOptions/option.js:
```diff
@@ -1,62 +1,85 @@
+    const optionType = getOptionType({ attribute_code, values });
+
+    return optionType === 'swatch' ? SwatchList : TileList;
+};
```
Note that I extracted this function from the component body. If it can exist statically, it should.

jimbo commented on packages/venia-concept/src/components/ProductOptions/option.js:
```diff
@@ -1,62 +1,85 @@
+    const ValueList = useMemo(() => getListComponent(attribute_code, values), [
+        attribute_code,
+        values
+    ]);
```
This is also a good place for `useMemo`.

dani97 replied:
Yes, Thanks for pointing that out.

jimbo commented on packages/venia-concept/src/components/ProductOptions/option.js:
```diff
@@ -1,62 +1,85 @@
+    ]);
+
+    const valuesMap = useMemo(() => {
+        return new Map(
```
Good use of `useMemo` here. 👍

dani97 replied:
Thanks. Credits to @sirugh. Now I have got understanding of useMemo hook  :)

jimbo commented on packages/venia-concept/src/components/ProductOptions/option.js:
```diff
@@ -1,62 +1,85 @@
+        selection => {
+            const [selectedValue] = Array.from(selection);
+
+            setSelection(valuesMap.get(selectedValue));
```
Note that we're now storing the selected label in state, rather than the set of indices.

jimbo commented on packages/venia-concept/src/components/ProductOptions/option.js:
```diff
@@ -1,62 +1,85 @@
+                items={values}
+                onSelectionChange={handleSelectionChange}
+            />
+            <p className={classes.selection}>{selectedValueLabel}</p>
```
Consider making this paragraph (`<p>`) a description list (`<dl>`) and splitting the label and value into term (`<dt>`) and description (`<dd>`).

sirugh commented:
@dani97 I am going to take over this PR since we haven't heard from you in a while. I'll be resolving the conflict files and addressing any remaining feedback.

dani97 commented:
Thanks @sirugh . I couldn't contribute for over a month. I hope I will join you soon :)
